ingestion-beam/src/main/java/com/mozilla/telemetry/io/Write.java

```diff
@@ -148,6 +166,98 @@ public FileOutput(ValueProvider<String> outputPrefix, OutputFileFormat format,
     }
   }
 
+  /**
+   * Implementation of writing to local or remote files.
+   *
+   * <p>For details of the intended behavior for file paths, see:
+   * https://github.com/mozilla/gcp-ingestion/tree/master/ingestion-beam#output-path-specification
+   */
+  public static class AvroOutput extends Write {
+
+    private final ValueProvider<String> outputPrefix;
+    private final Duration windowDuration;
+    private final ValueProvider<Integer> numShards;
+    private final Compression compression;
+    private final InputType inputType;
+    private final AvroSchemaStore schemaStore;
+    private final PubsubMessageRecordFormatter formatter = new PubsubMessageRecordFormatter();
+    private final IdentityRecordFormatter identityFormatter = new IdentityRecordFormatter();
+
+    /** Public constructor. */
+    public AvroOutput(ValueProvider<String> outputPrefix, Duration windowDuration,
+        ValueProvider<Integer> numShards, Compression compression, InputType inputType,
+        ValueProvider<String> schemasLocation) {
+      this.outputPrefix = outputPrefix;
+      this.windowDuration = windowDuration;
+      this.numShards = numShards;
+      this.compression = compression;
+      this.inputType = inputType;
+      this.schemaStore = AvroSchemaStore.of(schemasLocation);
+    }
+
+    @Override
+    public WithErrors.Result<PDone> expand(PCollection<PubsubMessage> input) {
+      final PCollectionView<AvroSchemaStore> schemaSideInput = input.getPipeline()
+          .apply(Create.of(schemaStore)).apply(View.asSingleton());
+
+      ValueProvider<DynamicPathTemplate> pathTemplate = NestedValueProvider.of(outputPrefix,
+          DynamicPathTemplate::new);
+      ValueProvider<String> staticPrefix = NestedValueProvider.of(pathTemplate,
+          value -> value.staticPrefix);
+
+      ParDo.SingleOutput<PubsubMessage, KV<Map<String, String>, GenericRecord>> intoGenericRecord = ParDo
```

Comment:
```
java.lang.IllegalStateException: Unable to return a default Coder for Write.AvroOutput/PubsubMessageToGenericRecord/ParMultiDo(Anonymous).output [PCollection]. Correct one of the following root causes:
  No Coder has been manually specified;  you may do so using .setCoder().
  Inferring a Coder from the CoderRegistry failed: Cannot provide coder for parameterized type org.apache.beam.sdk.values.KV<java.util.Map<java.lang.String, java.lang.String>, org.apache.avro.generic.GenericRecord>: Unable to provide a Coder for org.apache.avro.generic.GenericRecord.
  Building a Coder using a registered CoderProvider failed.
  See suppressed exceptions for detailed failures.
  Using the default output Coder from the producing PTransform failed: PTransform.getOutputCoder called.
```

Reply:
I hadn't totally thought through this, but looks like this is going to be a significant problem. It looks like AvroCoder requires a single schema be supplied. I don't think it's possible to store a PCollection containing GenericRecord of varying schemas. I believe AvroIO.Sink delays encoding to GenericRecord until after a destination has been determined, so from then on all records have the same schema and be encoded together in a single collection.

Looking at AvroIO's "Writing Data to Multiple Destinations" example, it looks like they define a custom DynamicAvroDestinations subclass that includes a hook for calling a method to encode to GenericRecord. That interface doesn't give us a way to bubble up exceptions to a separate collection.

Perhaps it's possible for us to hijack the interface to route to a failure destination if the message isn't encodable. We could have a special schema with just a single string field, and we could shove the payload in there. Our error output then would end up being avro-encoded. I'm not entirely sure this approach is feasible given the constraints of the interface.

Another possibility is that we maintain this custom ParDo, but only use it for a "dry run" of encoding to find records that are going to cause exceptions. We'd try encoding and if it passes, we'd pass through the unencoded payload string rather than the GenericRecord. Then, AvroIO.Sink would have to do the compute of encoding again, but we'd be guaranteed that it won't throw exceptions since we've already validated the avro encoding.

Another option would be to modify AvroIO itself to allow the exception handling we need and PR that code upstream. That's likely an ambitious task, though, so would only be feasible as a longer-term refactor rather than a solution for the purposes of getting this avro pipeline unblocked.

Reply:
I went with the first approach of stuffing the avro-encoded data into a Pubsub message. I think this is better than the second approach of a dry run, because writing and reading the binary data is probably faster than decoding the JSON payload twice.

I would be interested in seeing the options for adding error-handling closer to the FileIO writer. In a similar vein, I think another useful feature would be dynamic Parquet outputs, since I've already worked around the error handling in this PR.